xmltramp.py

```diff
@@ -126,105 +151,135 @@ def __str__(self):
             out = []
             for x in self._dir:
-                if isinstance(x, Element) and x._name == n: out.append(x)
+                if isinstance(x, Element) and x._name == n:
+                    out.append(x)
             return out
-        else: # d['foo'] == first <foo>
-            if self._dNS and not islst(n): n = (self._dNS, n)
+        else:  # d['foo'] == first <foo>
+            if self._dNS and not islst(n):
+                n = (self._dNS, n)
             for x in self._dir:
-                if isinstance(x, Element) and x._name == n: return x
+                if isinstance(x, Element) and x._name == n:
+                    return x
             raise KeyError(n)
 
     def __setitem__(self, n, v):
-        if isinstance(n, type(0)): # d[1]
+        if isinstance(n, type(0)):  # d[1]
             self._dir[n] = v
         elif isinstance(n, slice(0).__class__):
             # d['foo':] adds a new foo
             n = n.start
-            if self._dNS and not islst(n): n = (self._dNS, n)
+            if self._dNS and not islst(n):
+                n = (self._dNS, n)
 
             nv = Element(n)
             self._dir.append(nv)
 
-        else: # d["foo"] replaces first <foo> and dels rest
-            if self._dNS and not islst(n): n = (self._dNS, n)
+        else:  # d["foo"] replaces first <foo> and dels rest
+            if self._dNS and not islst(n):
+                n = (self._dNS, n)
 
-            nv = Element(n); nv._dir.append(v)
+            nv = Element(n)
+            nv._dir.append(v)
             replaced = False
 
             todel = []
-            for i in range(len(self)):
+            for i in xrange(len(self)):
                 if self[i]._name == n:
                     if replaced:
                         todel.append(i)
                     else:
                         self[i] = nv
                         replaced = True
-            if not replaced: self._dir.append(nv)
-            for i in todel: del self[i]
+            if not replaced:
+                self._dir.append(nv)
+            for i in sorted(todel, reverse=True):
+                del self[i]
 
+    # this was completely invalid
     def __delitem__(self, n):
-        if isinstance(n, type(0)): del self._dir[n]
+        if isinstance(n, type(0)):
+            del self._dir[n]
         elif isinstance(n, slice(0).__class__):
             # delete all <foo>s
             n = n.start
-            if self._dNS and not islst(n): n = (self._dNS, n)
+            if self._dNS and not islst(n):
+                n = (self._dNS, n)
 
-            for i in range(len(self)):
-                if self[i]._name == n: del self[i]
+            for i in reversed(range(len(self))):
+                if self[i]._name == n:
+                    del self[i]
         else:
             # delete first foo
             for i in range(len(self)):
-                if self[i]._name == n: del self[i]
-                break
```

Comment:
Nothing could be deleted, because the `break` command is outside `if` condition. The loop can terminate before the required item is found. This is now fixed.